### PR TITLE
websocket source

### PR DESCRIPTION
**Pull request by mikedewar**
this also contains the edits from #210 which should be merged first, and should be considered a req for this. 

this is a websocket source, with connect, receive, and send blocks.

### CODE REVIEW

**Comment by mikedewar** [comment]
also has some tidys the slack pattern required

**Comment by nikhan** [comment]
where did we land on what happens when the connection is lost?
does this drop messages?

**Comment by nikhan** [comment]
i merged 210 and now oops

**Comment by mikedewar** [comment]
:boom:

**Comment by nikhan** [comment]
squash

File: core/array_blocks.go [hunk 1]
```diff
@@ -60,7 +60,7 @@ func Last() Spec {
 				out[0] = NewError("zero length array passed to last")
 				return nil
 			}
-			out[0] = arr[len(arr)]
+			out[0] = arr[len(arr)-1]
```
Comment (nikhan) [inline core/array_blocks.go]: ??????????

File: core/membership.go [hunk 2]
```diff
@@ -58,12 +58,12 @@ func InString() Spec {
 		Kernel: func(in, out, internal MessageMap, s Source, i chan Interrupt) Interrupt {
 			substring, ok := in[0].(string)
 			if !ok {
-				out[0] = NewError("HasField requires strings")
+				out[0] = NewError("inString requires string for substring")
```
Comment (nikhan) [inline core/membership.go]: ????

File: core/membership.go [hunk 3]
```diff
@@ -58,12 +58,12 @@ func InString() Spec {
...
 				return nil
 			}
 			superstring, ok := in[1].(string)
 			if !ok {
-				out[0] = NewError("HasField requires strings")
+				out[0] = NewError("inString requires string for string")
```
Comment (nikhan) [inline core/membership.go]: ????

**Comment by mikedewar** [comment]
squahsed! Gimme some :heart:

**Comment by nikhan** [comment]
:dog: :horse:

**Pull request merged**